subprojects/version-control/src/main/java/org/gradle/vcs/git/internal/GitVersionRef.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.vcs.git.internal;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Splitter;
+import org.eclipse.jgit.lib.ObjectId;
+import org.eclipse.jgit.lib.Ref;
+import org.gradle.vcs.VersionRef;
+
+import java.util.List;
+
+public class GitVersionRef implements VersionRef {
+    private final String version;
+    private final String canonicalId;
+
+    private GitVersionRef(String version, String canonicalId) {
+        this.version = version;
+        this.canonicalId = canonicalId;
+    }
+
+    public static GitVersionRef from(Ref ref) {
+        ObjectId commitId = ref.getPeeledObjectId() == null ? ref.getObjectId() : ref.getPeeledObjectId();
+        return new GitVersionRef(extractName(ref), commitId.getName());
```

Comment:
❓ Is there anything else that would be useful when looking for problems?

Reply:
There are definitely a couple of other fields on Ref which we could log or something. But, I don't think we should make them part of the type. I'd rather add that logging when we know we need it.